working/macros/feature-specification.md

```diff
@@ -304,20 +304,69 @@ class, allowing you to build up your code fragments however you like.
 
 ## Scoping
 
-**TODO**: Fill in this section with more detail and a real proposal.
+### Resolved identifiers
 
 Macros will likely want to introduce references to identifiers that are not in
 the scope of the library in which they are running, but are in the scope of the
-macro itself.
+macro itself, or possibly even in some other library that is known to the
+macro, but not imported by the library where the macro is applied.
 
-We will want some way of providing an affordance to emit a reference to
-something from the macro scope, but in the code generated for the original
-library.
+Even if an identifier is expected to be in scope of the library in which the
+macro is running (lets say its exported by the macro library), that identifier
+could be shadowed by another identifier in the library.
 
-We do have the start of something like this already available in the `builder`
-api - these have APIs to get a reference to a `Type` object. We will want to add
-the ability to do the same for any arbitrary identifier, and then the ability
-to emit references to these inside of [Code][] objects.
+To enable a macro to safely emit a reference to a known identifier, there is
+a `ResolvedIdentifier` subtype of `Code`. This class takes both a simple name
+for the identifier (no prefix allowed), as well as a library uri, where that
+identifier should be looked up.
+
+The generated code should be equivalent to adding a new import to the library,
+with the specified uri and a unique prefix. In the code the identifier will be
+emitted with that unique prefix followed by its simple name.
+
+Note that technically this allows macros to add references to libraries that
+the macro itself does not depend on, and the users application also may not
+depend on. This is discouraged, but not prevented, and should result in an error
+if it happens.
```

Comment:
I'm not crazy about this. It adds a *lot* of power to let a macro freely compose a library URI and synthesize an identifier in its scope, and I don't see much justification for that. It means you could do potentially really weird stuff like have a macro application create an import cycle that wasn't there previously.

My initial pitch back when I wrote the Code API strawman is that there is a Scope object that you can use to create identifiers that resolve to that scope. Then there would be some way for a macro definition to get a Scope for the current scope where the macro is defined, and maybe one for where the macro is applied. I think that's probably all we need?

Reply:
Ooh, that's a really good point. I think that's going to require more thought and design. Maybe file a tracking issue so we can work through it outside of this PR?

Reply:
For this PR, do you think I should land it as is or replace the details with an TODO?

Reply:
How about leaving what you have but adding a TODO saying that it needs further thought?